sources/core/client.cpp

```diff
@@ -60,6 +60,11 @@ client::~client(void) {
     m_client.disconnect(true);
   }
 
+  // sparhawk@gmx.at
+  // If there are pending callbacks we have to wait until they are completed
+  // otherwise this will cause a memory corruption if they return after this point.
+  wait_for_completion();
```

Comment:
But if no callbacks are pending, then `notify_all()` will not be called from `connection_receive_handler` and `m_sync_condvar.wait` will hang.  So I don't think this works.

Reply:
I take that back.  However, there is another problem:  If there are uncommitted commands, then wait for completion will not return.

Reply:
@skeetor , have a look at this please https://github.com/steple/cpp_redis/commit/b69991470c4ea35b29279b79b8b836c0321ad315

This will pass all tests.  I'll try to add a test case that tests for the original bug of the pending callback.

Reply:
So does this mean that I can withdraw my PR and you will submit yours?

Reply:
I will, but my PR is missing a test case for the bug that you had found.  I'll need to add that before I can submit it.  If you have an idea on how to test it, please let me know.

Reply:
I will, but my PR is missing a test case for the bug that you had found. I'll need to add that before I can submit it. If you have an idea on how to test it, please let me know.

Reply:
I am unable to reproduce what I think is the bug you described.  Can you provide an example or a more detailed description?